xbmc/application/ApplicationPlayer.cpp

```diff
@@ -473,25 +472,34 @@ int CApplicationPlayer::GetSubtitle()
     return 0;
 }
 
-bool CApplicationPlayer::GetSubtitleVisible()
+bool CApplicationPlayer::GetSubtitleVisible() const
 {
-  std::shared_ptr<IPlayer> player = GetInternal();
-  return (player && player->GetSubtitleVisible());
+  const auto player = GetInternal();
+  return player && player->GetSubtitleVisible();
 }
 
-bool CApplicationPlayer::CanPause()
+bool CApplicationPlayer::CanPause() const
 {
-  std::shared_ptr<IPlayer> player = GetInternal();
+  const auto player = GetInternal();
   return (player && player->CanPause());
 }
 
+bool CApplicationPlayer::HasTeletextCache() const
+{
+  const auto player = GetInternal();
+  if (player)
+    return player->HasTeletextCache();
+  else
+    return false;
+}
+
 std::shared_ptr<TextCacheStruct_t> CApplicationPlayer::GetTeletextCache()
 {
   std::shared_ptr<IPlayer> player = GetInternal();
   if (player)
     return player->GetTeletextCache();
   else
-    return NULL;
+    return nullptr;
```

Comment:
new warning?

Reply:
To avoid going past stdlib and exposing raw pointers, you could return `std::shared_ptr<TextCacheStruct_t>{}`.

I also tend to like RVO:

```
--- a/xbmc/ApplicationPlayer.cpp
+++ b/xbmc/ApplicationPlayer.cpp
@@ -488,11 +488,13 @@ bool CApplicationPlayer::CanPause()
 
 std::shared_ptr<TextCacheStruct_t> CApplicationPlayer::GetTeletextCache()
 {
+  std::shared_ptr<TextCacheStruct_t> teletextCache;
+
   std::shared_ptr<IPlayer> player = GetInternal();
   if (player)
-    return player->GetTeletextCache();
-  else
-    return NULL;
+    teletextCache = player->GetTeletextCache();
+
+  return teletextCache;
 }
 
 float CApplicationPlayer::GetPercentage() const
```

Reply:
i only replaced NULL with the modern nullptr. I rather remove that hunk than start rewriting functions in this PR.

Reply:
if you prefer i can of course make it {}